micro: add documentation on porting ops from lite to micro

rkuester opened the pull request.

rkuester committed:
micro: add draft PORTING.md on porting ops from lite to micro

advaitjain commented on tensorflow/lite/micro/kernels/PORTING.md:
```diff
@@ -0,0 +1,251 @@
+for review and acceptance.  The process results in multiple pull requests, or
+PRs. Multiple, [small PRs][] are easier for the project to review and merge.
+
+[small PRs]: https://google.github.io/eng-practices/review/developer/small-cls.html
```
I didn't know we could do this in markdown, its very cool!

Can I suggest moving all of these links to the top of the file? And add a comment that you are defining links that will be reused throughout the file (for folks like me who don't know about this feature)? If I am reading the markdown itself from top to bottom, its easier if I know what all links have been defined.

rkuester replied:
Done.

I've never thought to put them at the top, but it looks fine. From what I see, people tend to think of reference-links as footnotes, and put them at the bottom. I sorta footnote-ed after each paragraph for proximity's sake, but I agree that turned out ugly.

8bitmp3 replied:
> I didn't know we could do this in markdown, its very cool!

+1 @advaitjain 🙂 cc @lamberta @MarkDaoust @anirudh161

advaitjain commented on tensorflow/lite/micro/kernels/PORTING.md:
```diff
@@ -0,0 +1,251 @@
+setup a development environment. This guide contains step-by-step instructions
+for the specific task of porting reference ops from Lite to Micro.
+
+[Micro Contributing Guidelines]: https://github.com/tensorflow/tensorflow/blob/master/tensorflow/lite/micro/CONTRIBUTING.md
```
move to the top

advaitjain commented on tensorflow/lite/micro/kernels/PORTING.md:
```diff
@@ -0,0 +1,251 @@
+issue. Link constituent PRs to this issue. See the article [Providing
+Context][] for background on documenting your work via bug reports.
+
+[Providing Context]: https://testing.googleblog.com/2017/09/code-health-providing-context-with.html
```
move to top

advaitjain commented on tensorflow/lite/micro/kernels/PORTING.md:
```diff
@@ -0,0 +1,251 @@
+    A simple example is [PR #45307][], and a more complicated example is [PR
+    #46021][].
+
+    [`ParseOpDataTfLite()`]: https://github.com/tensorflow/tensorflow/blob/d8394a6d774f5e3c02d97f1fc18ff445199db598/tensorflow/lite/core/api/flatbuffer_conversions.cc#L135
+    [PR #45307]: https://github.com/tensorflow/tensorflow/pull/45307
+    [PR #46021]: https://github.com/tensorflow/tensorflow/pull/46021
```
move to top?

advaitjain commented on tensorflow/lite/micro/kernels/PORTING.md:
```diff
@@ -0,0 +1,251 @@
+
+1.  Add the new header to the build in
+    *tensorflow/lite/kernels/internal/BUILD* under *reference_base* and
+    *legacy_reference_base*. E.g., [for FILL][].
+
+    [for FILL]: https://github.com/tensorflow/tensorflow/pull/45311/commits/92f459e6b917fa5099ef5317d14c5100d33a86f0#diff-0b0fc9e1affece3c5a141ee9326f882876b6b958bc8b12a7c01d7540dc04983e
```
since this link is only used once let's just make it [for FILL](link)

advaitjain commented on tensorflow/lite/micro/kernels/PORTING.md:
```diff
@@ -0,0 +1,251 @@
+    link to the GitHub issue created to document the port. A good example is
+    [PR #45647][].
+
+[#45306]: https://github.com/tensorflow/tensorflow/issues/45306
+[PR #45311]: https://github.com/tensorflow/tensorflow/pull/45311
+[PR #45457]: https://github.com/tensorflow/tensorflow/pull/45457
+[PR #45646]: https://github.com/tensorflow/tensorflow/pull/45646
+[PR #45647]: https://github.com/tensorflow/tensorflow/pull/45647
+[pre-submit checklist]: https://github.com/tensorflow/tensorflow/blob/master/tensorflow/lite/micro/CONTRIBUTING.md#before-submitting-your-pr
+[reference_ops.h]: https://github.com/tensorflow/tensorflow/blob/92f459e6b917fa5099ef5317d14c5100d33a86f0/tensorflow/lite/kernels/internal/reference/reference_ops.h
+[general porting guidelines]: #general-porting-guidelines
```
move to top?

advaitjain commented on tensorflow/lite/micro/kernels/PORTING.md:
```diff
@@ -0,0 +1,251 @@
+# Porting Reference Ops from Lite to Micro
```
add this line to the top to prevent internal google formatting tools from changing the file:

```
<!-- mdformat off(b/169948621#comment2) -->
```

advaitjain commented on tensorflow/lite/micro/kernels/PORTING.md:
```diff
@@ -0,0 +1,251 @@
+# Porting Reference Ops from Lite to Micro
+
```
might be worth a TOC using gh-md-toc similar to https://github.com/tensorflow/tensorflow/blob/faf9a3386c46fe1b18686e551abc21e4b2f10547/tensorflow/lite/micro/CONTRIBUTING.md#L3-L6

advaitjain commented on tensorflow/lite/micro/kernels/README.md:
```diff
@@ -0,0 +1,5 @@
+# Info
+
+*   [Porting Ops from Lite to Micro](./PORTING.md) explains, step-by-step, the
```
let's rename and move to docs/porting_reference_ops.md

porting.md is too generic since we also have docs on porting TFLM to a new target etc.

advaitjain reviewed (changes requested):
Thanks! Small nit comments.

Also, once https://github.com/tensorflow/tensorflow/pull/48807 is merged, you should be able to update this PR and have the TFLM CI build pass.

rkuester committed:
micro PORTING.md: move reference-style links to top of document

Per feedback on PR #48805, move the reference-style links to the
top of the document.

rkuester committed:
micro PORTING.md: prevent internal Google tools from munging

Add a header comment to the .md file that prevents internal
Google tools from changing the file.

rkuester committed:
micro PORTING.md: add TOC

Add a table of contents per feedback from review of PR #48805.

rkuester committed:
micro PORTING.md: move to docs/ and rename

Per review feedback from PR #48805, move PORTING.md to docs/ and
rename. *Porting* is too generic, because we also have docs on
porting to new target platforms.

8bitmp3 commented on tensorflow/lite/micro/docs/porting_reference_ops.md:
```diff
@@ -0,0 +1,267 @@
+
+1.  Use `clang-format` to make sure the code is properly formatted.
+
+        clang-format --style=google -i $(git ls-files -m | grep -E '\.cc|\.h')
```
Small suggestion - for command lines, you might consider using code fences (https://developers.google.com/style/code-syntax?hl=en)/code blocks (https://www.tensorflow.org/community/contribute/docs_style#write_about_code), specifying the "language" (e.g. `shell`, `python`, etc):

e.g.

```
    ```shell
    python3 awesome.py --style=awesome
    ```
````

rkuester replied:
I've done as you suggested in 5db4106, and further changed some formatting per the style guides you cited. Thanks for the review.

8bitmp3 replied:
No worries, glad I could help

rkuester committed:
micro: fixup porting reference ops doc per style guide

Cleanup the porting_reference_ops document according to review
suggestions and the Google and TF documentation style guides.

- Use fences with language tags on code blocks
- Use code spans instead of emphasis in serveral places

advaitjain commented:
> @rkuester can you please check sanity build failures ?

the build failures are unrelated to the current PR and can be ignored.
